cohttp-eio/src/server.ml

```diff
@@ -1,205 +1,113 @@
-  Buf_write.char writer ' ';
-  Buf_write.string writer status;
-  Buf_write.string writer "\r\n";
-  Rwer.write_headers writer headers;
-  Buf_write.string writer "\r\n";
-  let write_chunked_trailers =
-    Option.map Http.Request.supports_chunked_trailers request
-  in
-  Body.write_body ?write_chunked_trailers writer body
-
-(* request parsers *)
-
-let meth =
-  let open Eio.Buf_read.Syntax in
-  let+ meth = Rwer.(token <* space) in
-  Http.Method.of_string meth
-
-let resource =
-  let open Eio.Buf_read.Syntax in
-  Rwer.(take_while1 (fun c -> c != ' ') <* space)
-
-let[@warning "-3"] http_request t =
-  let open Eio.Buf_read.Syntax in
-  let meth = meth t in
-  let resource = resource t in
-  let version = Rwer.(version <* crlf) t in
-  let headers = Rwer.http_headers t in
-  let encoding = Http.Header.get_transfer_encoding headers in
-  { Http.Request.meth; resource; version; headers; scheme = None; encoding }
-
-(* main *)
-
-let rec handle_request clock client_addr reader writer flow handler =
-  match http_request reader with
-  | request ->
-      let response, body = handler (request, reader, client_addr) in
-      write_response ~request clock writer (response, body);
-      if Http.Request.is_keep_alive request then
-        handle_request clock client_addr reader writer flow handler
-  | (exception End_of_file)
-  | (exception Eio.Io (Eio.Net.E (Connection_reset _), _)) ->
-      ()
-  | exception (Failure _ as ex) ->
-      write_response clock writer bad_request_response;
-      raise ex
-  | exception ex ->
-      write_response clock writer internal_server_error_response;
-      raise ex
-
-let connection_handler (handler : handler) clock flow client_addr =
-  let reader = Buf_read.of_flow ~initial_size:0x1000 ~max_size:max_int flow in
-  Buf_write.with_flow flow (fun writer ->
-      handle_request clock client_addr reader writer flow handler)
-
-let run_domain env ssock handler =
-  let on_error exn =
-    Printf.fprintf stderr "Error handling connection: %s\n%!"
-      (Printexc.to_string exn)
-  in
-  let handler = connection_handler handler env#clock in
-  Switch.run (fun sw ->
-      let rec loop () =
-        Eio.Net.accept_fork ~sw ssock ~on_error handler;
-        loop ()
-      in
-      loop ())
-
-let run ?(socket_backlog = 128) ?(domains = 1) ~port env handler =
-  Switch.run @@ fun sw ->
-  let domain_mgr = Eio.Stdenv.domain_mgr env in
-  let ssock =
-    Eio.Net.listen (Eio.Stdenv.net env) ~sw ~reuse_addr:true ~reuse_port:true
-      ~backlog:socket_backlog
-      (`Tcp (Eio.Net.Ipaddr.V4.loopback, port))
-  in
-  for _ = 2 to domains do
-    Eio.Std.Fiber.fork ~sw (fun () ->
-        Eio.Domain_manager.run domain_mgr (fun () ->
-            run_domain env ssock handler))
-  done;
-  run_domain env ssock handler
-
-(* Basic handlers *)
-
-let not_found_handler _ = not_found_response
+  handle ()
+
+let run socket server =
+  Eio.Net.run_server socket ~on_error:raise (fun socket peer_address ->
+      let () =
```

Comment:
I suggest running a new switch in the `connection_handler` passed to `Eio.Net.run_server`:

```
(fun socket peer_address ->
  Eio.Switch.run (fun sw ->
  ...
))
```

The new switch is passed on to the user provided callback (see also suggested changes to `IO.conn`) and is used to group any fibers and resources allocated by the user. 

The `connection_handler` will only return when all resources are released. This also works nicely with the `max_connections:int` argument that can be passed to `Eio.Net.run_server` as the connection will be considered open as long as there are fibers running that were started by the user provided callback.